bpf/proxy/syncer.go

```diff
@@ -1423,6 +1444,27 @@ func ServicePortEqual(a, b k8sp.ServicePort) bool {
 		stringsEqual(a.TopologyKeys(), b.TopologyKeys())
 }
 
+func ServiceEpsEqual(a, b []k8sp.Endpoint) bool {
```

Comment:
I think it's not uncommon to have a large number of endpoints (e.g. due to having a prometheus endpoint in every pod) so I don't think we can have a `O(n^2)` algorithm here.  Can we assume the slices are sorted?

If not, assuming Endpoint is hashable, you could do 
```
if len(a) != len(b) {
    return false
}
if len(a) > someThresh {
    eps  := make(map[k8sp.Endpoint]struct{}, len(a))
    for _, aa := range a {
        eps[aa] = struct{}{}
    }
    equal := true
    for _, bb := range b {
        _, ok := eps[bb]
       if !ok {
           return false
       }
    }
    return true
}
// else use the n^2 algo
```
If the input is usually in the same order but it's not guaranteed, you could speed up the n^2 to n in the happy case by doing:
```
for aIdx, aa := range a {
	found := false
	for bIdx := range b {
		bb = (aIdx + bIdx) % len(b)
		if aa.Equal(bb) {
			found = true
			break
		}
	}
	if !found {
		return false
	}
}
```

Reply:
I opted for the n^2 because I have no idea whether it is ordered in any way. But, yes, it may be an issue. I will have a look at it. Given the fact that we test for "no change" in the service rather than generic equality, I think we can assume that the order is preserved and a simple linear scan is good enough.